File: src/brs/grpc/handlers/CompleteBasicTransactionHandler.java
package brs.grpc.handlers;

import brs.Attachment;
import brs.grpc.GrpcApiHandler;
import brs.grpc.proto.ApiException;
import brs.grpc.proto.BrsApi;
import brs.services.TimeService;
import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;

public class CompleteBasicTransactionHandler implements GrpcApiHandler<BrsApi.BasicTransaction, BrsApi.BasicTransaction> {

    private final TimeService timeService;

    public CompleteBasicTransactionHandler(TimeService timeService) {
        this.timeService = timeService;
    }

    @Override
    public BrsApi.BasicTransaction handleRequest(BrsApi.BasicTransaction basicTransaction) throws Exception {
        try {
            BrsApi.BasicTransaction.Builder builder = basicTransaction.toBuilder();
            Attachment.AbstractAttachment attachment = Attachment.AbstractAttachment.parseProtobufMessage(basicTransaction.getAttachment());
            builder.setVersion(1); // TODO
            builder.setType(attachment.getTransactionType().getType());
            builder.setSubtype(attachment.getTransactionType().getSubtype());
            builder.setTimestamp(timeService.getEpochTime());
            if (builder.getAttachment().equals(Any.getDefaultInstance())) {
                builder.setAttachment(Attachment.ORDINARY_PAYMENT.getProtobufMessage());
            }
            return builder.build();
        } catch (InvalidProtocolBufferException e) {
            throw new ApiException("Could not parse an Any: " + e.getMessage());
        }
    }
}

File: src/main/java/brs/grpc/handlers/CompleteBasicTransactionHandler.kt
package brs.grpc.handlers

import brs.Attachment
import brs.DependencyProvider
import brs.grpc.GrpcApiHandler
import brs.grpc.proto.ApiException
import brs.grpc.proto.BrsApi
import burst.kit.crypto.BurstCrypto
import com.google.protobuf.Any
import com.google.protobuf.InvalidProtocolBufferException

class CompleteBasicTransactionHandler(private val dp: DependencyProvider) : GrpcApiHandler<BrsApi.BasicTransaction, BrsApi.BasicTransaction> {
    override suspend fun handleRequest(request: BrsApi.BasicTransaction): BrsApi.BasicTransaction {
        try {
            val builder = request.toBuilder()
            val attachment = Attachment.AbstractAttachment.parseProtobufMessage(dp, request.attachment)
            if (builder.deadline == 0) {
                builder.deadline = 1440
            }
            if (builder.senderId == 0L) {
                builder.senderId = BurstCrypto.getInstance().getBurstAddressFromPublic(builder.senderPublicKey.toByteArray()).burstID.signedLongId
            }
            builder.version = dp.transactionProcessor.getTransactionVersion(dp.blockchain.height)
            builder.type = attachment.transactionType.type.toInt()
            builder.subtype = attachment.transactionType.subtype.toInt()
            builder.timestamp = dp.timeService.epochTime
            if (builder.attachment == Any.getDefaultInstance()) {
                builder.attachment = Attachment.OrdinaryPayment(dp).protobufMessage
            }
            return builder.build()
        } catch (e: InvalidProtocolBufferException) {
            throw ApiException("Could not parse an Any: " + e.message)
        }
    }
}